text/0000-io-fs-2.1.md

```diff
@@ -0,0 +1,558 @@
+///
+/// Note that relative paths given to this function will use the current working
+/// directory as a base, and the current working directory is not managed in a
+/// thread-local fashion, so this function may need to be synchronized with
+/// other calls to `env::change_dir`.
+pub fn canonicalize<P: AsRef<Path>>(p: P) -> io::Result<PathBuf>;
+```
+
+## Tweaking `PathExt`
+
+Currently the `PathExt` trait is unstable, yet it is quite convenient! The main
+motivation for its `#[unstable]` tag is that it is unclear how much
+functionality should be on `PathExt` versus the `std::fs` module itself.
+Currently a small subset of functionality is offered, but it is unclear what the
+guiding principle for the contents of this trait are.
+
+This RFC proposes a few guiding principles for this trait:
+
+* Only read-only operations in `std::fs` will be exposed on `PathExt`. All
+  operations which require modifications to the filesystem will require calling
+  methods through `std::fs` itself.
+
+* Some inspection methods on `Metadata` will be exposed on `PathExt`, but only
+  those where it logically makes sense for `Path` to be the `self` receiver. For
+  example `PathExt::len` will not exist (size of the file), but
+  `PathExt::is_dir` will exist.
+
+Concretely, the `PathExt` trait will be expanded to:
+
+```rust
+pub trait PathExt {
+    fn exists(&self) -> bool;
+    fn is_dir(&self) -> bool;
+    fn is_file(&self) -> bool;
+    fn metadata(&self) -> io::Result<Metadata>;
+    fn symlink_metadata(&self) -> io::Result<Metadata>;
+    fn canonicalize(&self) -> io::Result<PathBuf>;
+    fn read_link(&self) -> io::Result<PathBuf>;
+    fn read_dir(&self) -> io::Result<ReadDir>;
+}
+
+impl PathExt for Path { ... }
+```
+
+## Expanding `DirEntry`
+
+Currently the `DirEntry` API is quite minimalistic, exposing very few of the
+underlying attributes. Platforms like Windows actually contain an entire
+`Metadata` inside of a `DirEntry`, enabling much more efficient walking of
+directories in some situations.
+
+The following APIs will be added to `DirEntry`:
+
+```rust
+impl DirEntry {
+    /// This function will return the filesystem metadata for this directory
+    /// entry. This is equivalent to calling `fs::symlink_metadata` on the
+    /// path returned.
+    ///
+    /// On Windows this function will always return `Ok` and will not issue a
+    /// system call, but on unix this will always issue a call to `stat` to
+    /// return metadata.
+    pub fn metadata(&self) -> io::Result<Metadata>;
+
+    /// Return what file type this `DirEntry` contains.
+    ///
+    /// On some platforms this may not require reading the metadata of the
+    /// underlying file from the filesystem, but on other platforms it may be
+    /// required to do so.
+    pub fn file_type(&self) -> io::Result<FileType>;
+
+    /// Returns the file name for this directory entry.
+    pub fn file_name(&self) -> OsString;
```

Comment:
What append if this is a directory or something else ? Do we call this function ?